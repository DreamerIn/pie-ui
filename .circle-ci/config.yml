# Javascript Node CircleCI 2.0 configuration file
#
# Check https://circleci.com/docs/2.0/language-javascript/ for more details
#
version: 2

defaults: &defaults
  working_directory: ~/repo
  docker:
    - image: circleci/node:11.10.0

jobs:
  build:
    <<: *defaults
    steps:
      - checkout

      - one:
        name: one
        command: echo one
#       # Download and cache dependencies
#       - restore_cache:
#           keys:
#             - v1-dependencies-{{ checksum "package.json" }}
#             # fallback to using the latest cache if no exact match is found
#             - v1-dependencies-

#       - run: yarn install

#       - save_cache:
#           paths:
#             - node_modules
#           key: v1-dependencies-{{ checksum "package.json" }}

#       # run tests!
#       - run:
#           name: Running Tests
#           command: yarn test

#       - run:
#           name: Running Lint
#           command: yarn lint

#   releaseNext:
#     <<: *defaults
#     branches:
#       only:
#         - develop
#     steps:
#       - attach_workspace:
#           at: ~/repo
#       - run:
#           name: Authenticate with registry
#           command: echo "//registry.npmjs.org/:_authToken=$npm_TOKEN" > ~/repo/.npmrc
#       - run:
#           name: Release Next
#           command: npm run release:next

# workflows:
#   version: 2
#   nightly:
#     triggers:
#       - schedule:
#           cron: '0 0 * * *'
#           filters:
#             branches:
#               only:
#                 - develop
#     jobs:
#       - releaseNext
