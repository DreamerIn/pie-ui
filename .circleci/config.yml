version: 2

defaults: &defaults
  working_directory: ~/repo
  docker:
    - image: circleci/node:11.10.0

reg_auth: &reg_auth
  name: Authenticate with registry
  command: echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" > ~/repo/.npmrc

restore_cache: &restore_cache
  keys:
    - v1-dependencies-{{ checksum "package.json" }}
    # fallback to using the latest cache if no exact match is found
    - v1-dependencies-

save_cache: &save_cache
  paths:
    - node_modules
  key: v1-dependencies-{{ checksum "package.json" }}

jobs:
  build:
    <<: *defaults

    steps:
      - checkout
      - run: *reg_auth
      - restore_cache: *restore_cache
      - run: yarn install
      - save_cache: *save_cache
      - run: scripts/build release --next --skipForcePublish

  deploy_demo:
    <<: *defaults

    steps:
      - checkout
      - run: *reg_auth
      - restore_cache: *restore_cache
      - run: yarn install
      - save_cache: *save_cache
      - run: scripts/build build buildDemo deployDemo --next

workflows:
  version: 2
  build:
    jobs:
      - build:
          filters:
            branches:
              only:
                - develop

  nightly:
    triggers:
      - schedule:
          cron: '0 0 * * *'
          filters:
            branches:
              only:
                - develop
    jobs:
      - deploy_demo
