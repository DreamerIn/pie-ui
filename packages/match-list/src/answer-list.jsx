import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import isEmpty from 'lodash/isEmpty';
import DragAndDropAnswer from './answer';

export class AnswerList extends React.Component {
  static propTypes = {
    classes: PropTypes.object,
    session: PropTypes.object.isRequired,
    showCorrect: PropTypes.bool.isRequired,
    disabled: PropTypes.bool.isRequired,
    onSessionChange: PropTypes.func,
    onRemoveAnswer: PropTypes.func,
    placeAnswer: PropTypes.func.isRequired,
    instanceId: PropTypes.string.isRequired,
    model: PropTypes.object.isRequired,
    prompt: PropTypes.string
  };

  getAnswerFromSession(index) {
    const { model, session, showCorrect } = this.props;
    const { config } = model;

    const answerId = showCorrect ? config.prompts[index].relatedAnswer : (session.value && session.value[index]);
    const answer = config.answers.find(answer => answer.id === answerId);

    return answer || {};
  }

  getCorrectOrIncorrectArray() {
    const { model, session, showCorrect } = this.props;
    const { config } = model;
    const sessionValue = session.value || new Array(config.prompts.length).fill(undefined);

    if (model.mode !== 'evaluate') {
      return [];
    }

    if (showCorrect) {
      return new Array(config.prompts.length).fill(true);
    }

    const correctPromptMap = config.prompts.reduce((obj, prompt) => {
      if (prompt.relatedAnswer) {
        obj[prompt.id] = prompt.relatedAnswer;
      }

      return obj;
    }, {});

    return sessionValue.map((val, index) => {
      const currentPrompt = config.prompts[index];

      return correctPromptMap[currentPrompt.id] === val;
    });
  }

  render() {
    const {
      classes,
      disabled,
      placeAnswer,
      instanceId,
      onRemoveAnswer,
      model
    } = this.props;
    const { config } = model;
    const correctnesArray = this.getCorrectOrIncorrectArray();

    return (
      <div className={classes.itemList}>
        {
          config.prompts.map((answer, index) => {
            const sessionAnswer = this.getAnswerFromSession(index);

            return (
              <DragAndDropAnswer
                key={index}
                index={index}
                correct={correctnesArray[index]}
                draggable={!isEmpty(sessionAnswer)}
                disabled={disabled}
                instanceId={instanceId}
                id={sessionAnswer.id}
                placeAnswer={(id, index) => placeAnswer(id, index)}
                title={sessionAnswer.title}
                type={'target'}
                onRemoveChoice={() => onRemoveAnswer(index)}
              />
            );
          })
        }
      </div>
    );
  }
}

const styles = () => ({
  itemList: {
    alignItems: 'flex-end',
    display: 'flex',
    flex: 1,
    flexDirection: 'column',
    justifyContent: 'space-between'
  }
});

export default withStyles(styles)(AnswerList);