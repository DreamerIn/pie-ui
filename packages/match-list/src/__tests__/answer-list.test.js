import * as React from 'react';
import { shallow } from 'enzyme';
import isArray from 'lodash/isArray';
import { AnswerList } from '../answer-list';
import { model, answer } from '../../demo/config';

describe('AnswerList', () => {
  const defaultProps = {
    model: model('1'),
    session: {},
    classes: {}
  };

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<AnswerList {...defaultProps} />);
  });

  describe('render', () => {
    it('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('logic', () => {
    describe('getCorrectOrIncorrectArray', () => {
      const mkTestForFn = (title, extraProps, val) => {
        it(title, () => {
          wrapper.setProps({
            ...defaultProps,
            ...extraProps
          });

          const value = wrapper.instance().getCorrectOrIncorrectArray();

          expect(isArray(value)).toBe(true);
          expect(value).toEqual(val);
        });
      };

      mkTestForFn(
        'returns an empty array for modes different than evaluate',
        {
          model: model('1', {
            mode: 'gather'
          })
        },
        []
      );

      mkTestForFn(
        'returns an empty array for modes different than evaluate',
        {
          model: model('1', {
            mode: 'gather'
          })
        },
        []
      );

      mkTestForFn(
        'returns an array appropriate values when mode is evaluate',
        {
          model: model('1', {
            mode: 'evaluate'
          })
        },
        [false, false, false, false]
      );

      mkTestForFn(
        'returns an array with appropriate values',
        {
          model: model('1', {
            mode: 'evaluate'
          }),
          session: {
            value: [1, 3, 4, 2]
          }
        },
        [true, true, true, true]
      );

      mkTestForFn(
        'returns an array with true as value for each index when showCorrect is true',
        {
          model: model('1', {
            mode: 'evaluate'
          }),
          showCorrect: true
        },
        [true, true, true, true]
      );

    });

    describe('getAnswerFromSession', () => {
      const mkTestForFn = (title, extraProps, index, val) => {
        const indexes = isArray(index) ? index : [index];
        const values = isArray(val) ? val : [val];

        it(title, () => {
          wrapper.setProps({
            ...defaultProps,
            ...extraProps
          });

          indexes.forEach((i, key) => {
            const value = wrapper.instance().getAnswerFromSession(i);

            expect(value).toEqual(values[key]);
          });
        });
      };

      mkTestForFn(
        'returns an empty object for an empty session',
        {
          model: model('1', {
            mode: 'gather'
          })
        },
        0,
        {}
      );

      mkTestForFn(
        'returns an appropriate answer for filled session',
        {
          model: model('1', {
            mode: 'gather'
          }),
          session: { value: [2, 1, 4, 3] }
        },
        0,
        { id: 2, title: "Answer 2" }
      );

      mkTestForFn(
        'returns all correct values when showCorrect is true',
        {
          model: model('1', {
            mode: 'gather',
            showCorrect: true
          }),
          session: { value: [1, 3, 4, 2] }
        },
        [0, 1, 2, 3],
        [answer(1), answer(3), answer(4), answer(2)]
      );

    });
  });
});
