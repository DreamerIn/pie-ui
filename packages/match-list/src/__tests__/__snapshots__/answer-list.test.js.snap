// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnswerList render renders correctly 1`] = `
<div>
  <DragSource(DropTarget(WithStyles(Answer)))
    draggable={false}
    index={0}
    key="0"
    onPlaceAnswer={[Function]}
    onRemoveChoice={[Function]}
    type="target"
  />
  <DragSource(DropTarget(WithStyles(Answer)))
    draggable={false}
    index={1}
    key="1"
    onPlaceAnswer={[Function]}
    onRemoveChoice={[Function]}
    type="target"
  />
  <DragSource(DropTarget(WithStyles(Answer)))
    draggable={false}
    index={2}
    key="2"
    onPlaceAnswer={[Function]}
    onRemoveChoice={[Function]}
    type="target"
  />
  <DragSource(DropTarget(WithStyles(Answer)))
    draggable={false}
    index={3}
    key="3"
    onPlaceAnswer={[Function]}
    onRemoveChoice={[Function]}
    type="target"
  />
</div>
`;
